### ISSUE TITLE
Improve startup performance

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi#3671]
The shortest possible `pulumi preview` or `pulumi up` is currently around 4 seconds.  We can likely improve this significantly

A typical preview using TypeScript and AWS looks as follows:
* **750ms-1000ms**: talking to Pulumi Service
* **500ms**: booting NodeJS
* **500ms**: initializing Pulumi inside NodeJS (including TS compilation, but I recall V8 hooks stuff was actually a bigger part of this)
* **2500ms**: configuring AWS provider.

<img width="960" alt="Screen Shot 2019-12-16 at 9 17 49 AM" src="https://user-images.githubusercontent.com/223467/70956195-203ef900-2028-11ea-8c3c-0c1de06b0bf3.png">
<img width="974" alt="Screen Shot 2019-12-16 at 9 02 20 AM" src="https://user-images.githubusercontent.com/223467/70956196-20d78f80-2028-11ea-925e-7eb4c4abbac2.png">

There are some cases where things are overall much worse - on a fresh tethered network connection and without having run Pulumi recently I saw overall worse performance across many of these steps, including much longer waits on network calls.  On more limited machines or networks, these results may be more representative, and may lead to further optimization opportunities.

<img width="1345" alt="Screen Shot 2019-12-16 at 9 01 30 AM" src="https://user-images.githubusercontent.com/223467/70956194-203ef900-2028-11ea-838e-5a6ee7ee7984.png">

From the above - some potential opportunities:
- [ ] Optimize provider configuration (this is mostly inherited from the upstream providers for AWS/Azure/GCP, but we could contribute general performance improvements - like avoiding network calls that are made multiple times).
- [ ] Attempt to improve NodeJS process launch
- [ ] Attempt to improve Pulumi NodeJS bootstrap (like v8hooks load, potentially moving it off critical path)
- [ ] Move some of the network calls to the Pulumi service backend off the critical path (run concurrently with other work)
- [ ] Improve `api/getStack` performance in the Pulumi service backend

**Comment by pgavlin** [comment]
> Move some of the network calls to the Pulumi service backend off the critical path (run concurrently with other work)

Which calls were you thinking of here?

**Comment by lukehoban** [comment]
> > Move some of the network calls to the Pulumi service backend off the critical path (run concurrently with other work)

> Which calls were you thinking of here?

The 5 calls that happen synchronously at the beginning of every preview/update:
- `api/getStack`
- `api/createUpdate`
- `api/startUpdate`
- `api/renewLease`
- `api/exportStack`

It isn't immediately clear any of these can be moved off the critical path - but (a) could we do fewer roundtrips on the critical path?  (b) Are all of these calls strictly necessary?

**Comment by pgavlin** [comment]
> but (a) could we do fewer roundtrips on the critical path? (b) Are all of these calls strictly necessary?

We could probably fold all of those calls together into a single `createAndStartUpdate` API on the service side.

**Comment by joeduffy** [comment]
Another data point, I was able to shave off 5 seconds of runtime off [this project](https://github.com/pulumi/customer-engineering/tree/master/demos/aws-ts-webserver) by pre-compiling a project (using `tsc`) and commenting out [these lines](https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/cmd/run/run.ts#L154-L163), which run auto-compilation using `node-ts`.

This is a surprising amount of time. Are there any smarts we can introduce around lazy compilation, like not doing it if the files didn't change since last time? I assumed `node-ts` would be smart enough to do incremental compilation, but either we're not using it in a way that is conducive to doing so, or it doesn't.

**Comment by joeduffy** [comment]
Also note that on very slow networks, https://github.com/pulumi/pulumi-aws/issues/873 seems to be a significant contributing factor to the performance issues.

**Comment by lukehoban** [comment]
A few more observations looking at some detailed logs of the initialization sequence:
- On first ever update of a stack, we `Configure` the AWS provider once for preview and once for update. On every future update, we `Configure` the AWS provider twice for preview and once for update (!).
- During the `Configure` for AWS we hit the EC2 metadata endpoint twice (once due to our own `PreConfigureCallback`, the other part of the upstream provider `Configure`).  Despite being set to timeout after `100ms`, each call reliably takes 1-1.5s.
- After those two calls, we then call `sts/GetCallerIdentity` twice (the upstream provider just happens to do this by default) and `ec2/DescribeAccountAttributes` once - each call taking .5-1s.  

Total of above is that `Configure` takes 3.5-6s.  And we call it twice during preview.

**Comment by pgavlin** [comment]
> we call it twice during preview.

Do you have a trace for this?

**Comment by lukehoban** [comment]
When this issue was opened - performance was *significantly* better than it is currently.  Today - even in the most trivial Pulumi Typescript program - about 4.5s of startup time is reliably spent compiling the user program.

It turns out that's due to https://github.com/pulumi/pulumi/pull/3627 which upgraded our `ts-node` version (coincidentally - merged the day after this issue was opened).  There was a *massive* regression in `ts-node` performance when they disabled caching.  And caching is not (yet) available again in `ts-node` - though it is tracked in https://github.com/TypeStrong/ts-node/issues/951 and related issues.

We will almost certainly have to revert back to an old `ts-node` and find a way to allow that to still use a new version of `typescript`.

Here are traces with latest `ts-node`:

<img width="546" alt="Screen Shot 2020-02-28 at 5 52 37 PM" src="https://user-images.githubusercontent.com/223467/75598776-39feb280-5a53-11ea-8928-3e3d45f46a62.png">

And with `7.0.0` of `ts-node`:

<img width="466" alt="Screen Shot 2020-02-28 at 5 51 54 PM" src="https://user-images.githubusercontent.com/223467/75598768-2f441d80-5a53-11ea-8e19-cbcbed005950.png">

**Comment by lukehoban** [comment]
The changes in #4004 and #4007 have reduced the time for a `pulumi up --yes` in steady state (no changes) on a the default AWS template (single S3 bucket) by >50%.

Run | Before | After
-- | -- | --
1 | 18.715 | 8.13
2 | 21.911 | 8.901
3 | 17.836 | 7.791
4 | 19.405 | 7.549
5 | 17.19 | 9.158
6 | 18.355 | 8.991
7 | 17.071 | 8.839
8 | 16.028 | 9.261
9 | 17.205 | 7.247
10 | 17.699 | 8.949
Average | 18.1415 | 8.4816

Of the ~8.5s about 7s of that is calls to `Configure` (one each fo the `preview` and `update` parts of the operation above).  Improving those is tracked in https://github.com/pulumi/pulumi-aws/issues/873 and is specific to the AWS provider.  Because we have moved `Configure` to run concurrently with the user process spawning, there is (at least in stable state) no long-pole cost to the Node.js process startup (it completes in ~1s, which is always faster than `Configure`).

The remaining 1.5s is roughly (each of these repeated twice):
- 400ms: Talking to api.pulumi.com
- 250ms: GetRequiredPlugins

(along with <200ms of "other")

I consider those "okay" for now.  So in the near-term, we will focus on https://github.com/pulumi/pulumi-aws/issues/873.

**Comment by gsuess** [comment]
One easy way to improve performance would be to remove `GeRequiredPlugins` from the loop altogether: #4607

**Comment by lukehoban** [comment]
I took another look at what long pole performance is like for a `pulumi up --skip-preview` of a stack with a single S3 bucket, in stable state (no changes to make).  I'm currently in Australia, which means the app.pulumi.com API calls are substantially higher latency.

I consistently see this no-op update take 4.0-4.2s (if I haven't run it in awhile - it takes 5-6s - more on that below).

Of that time:
* 2300ms of long pole is spent talking to api.pulumi.com before and after the update actually happens
* 1200ms of long pole in Configure 
* 300ms on a per-resource `PATCH` to api.pulumi.com
* ~250ms unaccounted for after these

<img width="1054" alt="Screen Shot 2021-02-18 at 3 31 26 PM" src="https://user-images.githubusercontent.com/223467/108305786-e1459e00-71fe-11eb-94c7-9a61f1e95b9e.png">

<img width="381" alt="Screen Shot 2021-02-18 at 3 39 12 PM" src="https://user-images.githubusercontent.com/223467/108306053-78125a80-71ff-11eb-9efb-8c834dde4d14.png">

The `PATCH` is especially concerning - this update was a no-op with only "sames", and should not have required any update to the state file.  I expect this is the same underlying issue causing https://github.com/pulumi/pulumi/issues/6197 (which will be much more pronounced with hundreds of resources and with a filestate backend that can be much slower).

The 2300ms of time talking to api.pulumi.com is more pronounced by my being in Australia - but is a *lot* of overhead to pay.  The latency of 7 round trips to the api server definitely feels more than is necessary.

The 1200ms in `Configure` is surprising since this is after having fixed https://github.com/pulumi/pulumi-aws/issues/873.  I'll need to look more deeply into this piece as a follow-up.

**Comment by lukehoban** [comment]
> The 1200ms in Configure is surprising since this is after having fixed pulumi/pulumi-aws#873. I'll need to look more deeply into this piece as a follow-up.

This appears to be nearly entirely due to a call to IAM GetUser which takes 1100ms.  That may be unavoidable for AWS.

**Comment by zyv** [comment]
We are using lots of all kinds of AWS resources, so maybe our terrible performance in #6197 is actually in large part related to `pulumi-aws`? I have to say though, that on a small project using exclusively `pulumi-github` I experience the same excruciating 10-15 minutes waits on `pulumi up`... totally crazy.

**Comment by lukehoban** [comment]
> the same excruciating 10-15 minutes waits on pulumi up.

What backend are you using?  Is it the S3 backend?  If so - I do expect this is related to #6197.  If you are able to reach out on the Pulumi Community Slack (I'm @Luke) - I'd love to get some additional details from you to help diagnose this.  10-15 minutes is definitely very far from expected!

**Comment by sam-goodwin** [comment]
I'm seeing extremely poor performance on a stack that simply creates an IAM Role. My CPU hangs on 100% CPU and the Pulumi Preview page just spins with no information for XYZ minutes

Original application was a Pulumi CDK one, but I was surprised by how poor the performance was so I dialed it back to the below app. I also pointed my `main` in `package.json` at `lib/index.js` instead of `src/index.ts` and saw no performance gains.

Not sure what's going on. 

FYI - I'm on a Mac M1

```ts
import { iam } from "@pulumi/aws";

const role = new iam.Role("my-role", {
  assumeRolePolicy: {
    Version: "2012-10-17",
    Statement: [
      {
        Effect: "Allow",
        Action: "sts:AssumeRole",
        Principal: {
          AWS: "lambda.amazonaws.com",
        },
      },
    ],
  },
});
```

**Comment by lukehoban** [comment]
@sam-goodwin that does not sound right. Any chance you could follow https://www.pulumi.com/docs/support/troubleshooting/#verbose-logging to collect diagnostic logs and see what is hanging?  What processes is hitting 100% CPU?

Additionally - sharing `pulumi about` may help. 

It might make sense to open a new issue as what you are reporting sounds different than the other issues discussed above in this issue.

**Comment by sam-goodwin** [comment]
Thanks @lukehoban - I opened up a new Issue to track this: https://github.com/pulumi/pulumi/issues/11946
